buildkite/bazel_auto_sheriff.py

```diff
@@ -0,0 +1,643 @@
+        self.main_result["commit"] = main_build_info["commit"]
+        self.main_result["build_number"] = main_build_info["number"]
+        job_infos = filter(lambda x: x != None, [extract_job_info_by_key(job) for job in main_build_info["jobs"]])
+        self.main_result["tasks"] = group_job_info_by_task(job_infos)
+        self.main_result["state"] = get_project_state(self.main_result["tasks"])
+
+        last_green_commit_url = bazelci.bazelci_last_green_commit_url(
+            bazelci.DOWNSTREAM_PROJECTS[self.project]["git_repository"], self.pipeline
+        )
+        self.main_result["last_green_commit"] = bazelci.get_last_green_commit(last_green_commit_url)
+
+
+    # Log all succeeded projects in the same annotate block
+    def __log_success(self, text):
+        with BuildInfoAnalyzer.success_log_lock:
+            BuildInfoAnalyzer.success_log.append(COLORS["HEADER"] + f"Analyzing {self.project}: " + COLORS["ENDC"] + COLORS["PASSED"] + text + COLORS["ENDC"])
+            info = [
+                "<details><summary><strong>:bk-status-passed: Success</strong></summary><p>",
+                "",
+                "```term",
+            ] + BuildInfoAnalyzer.success_log + [
+                "```",
+                "",
+                "</p></details>"
+            ]
+            print_info("success-info", "success", info, append = False)
+
+    # Log broken projects in their separate annotate block
+    def __log(self, c, text):
+        self.analyze_log.append(COLORS[c] + text + COLORS["ENDC"])
+        info = [
+            f"<details><summary><strong>:bk-status-failed: {self.project}</strong></summary><p>",
+            "",
+            "```term",
+        ] + self.analyze_log + [
+            "```",
+            "",
+            "</p></details>"
+        ]
+        print_info(self.pipeline, "warning", info, append = False)
+
+
+    # Implement the log function so that it can be called from BuildkiteClient
+    def log(self, text):
+        # Be smart here to reduce log length.
+        if self.analyze_log[-1].startswith(COLORS["INFO"] + "Waiting for "):
+            self.analyze_log.pop(-1)
+        self.__log("INFO", text)
+
+
+    def __trigger_bisect(self, tasks):
+        env = {
+            "PROJECT_NAME": self.project,
+            "TASK_NAME_LIST": ",".join(tasks) if tasks else "",
+        }
+        return CULPRIT_FINDER_PIPELINE_CLIENT.trigger_new_build("HEAD", f"Bisecting {self.project}", env)
+
+
+    # Search for certain message in the log to determine the bisect result.
+    # Return values are
+    #   1. A bisect result message
+    #   2. The commit of the culprit if found, otherwise None
+    def __determine_bisect_result(self, job):
+        bisect_log = CULPRIT_FINDER_PIPELINE_CLIENT.get_build_log(job)
+        pos = bisect_log.rfind("first bad commit is ")
+        if pos != -1:
+            start = pos + len("first bad commit is ")
+            culprit_commit = bisect_log[start:start + 40]
```

Comment:
where does the 40 come from?

Reply:
The length of a full commit hash is 40
Comment added.